File: go/500.600/541.ReverseString2.go
package main

import (
	"fmt"
)

func reverseStr(s string, k int) string {
	byts := []byte(s)
	n := len(byts)
	p, q := 0, k
	for p < n {
		if q > n {
			q = n
		}
		m := q - 1
		for p < m {
			byts[p], byts[m] = byts[m], byts[p]
			p++
			m--
		}
		p = q + k
		q += 2 * k
	}
	return string(byts)
}

func main() {
	fmt.Println(reverseStr("abcdefg", 2), "bacdfeg")
	fmt.Println(reverseStr("abcd", 2), "bacd")
}

File: 500.600/541.ReverseString2.cpp
#include <iostream>
#include <string>
using namespace std;

class ReverseString2
{
public:
    string reverseStr(string s, int k)
    {
        int n = s.size();
        int p = 0, q = k;
        while (p < n)
        {
            if (q > n)
            {
                q = n;
            }
            int m = q - 1;
            while (p < m)
            {
                char c = s[m];
                s[m] = s[p];
                s[p] = c;
                ++p;
                --m;
            }

            p = q + k;
            q += 2 * k;
        }
        return s;
    }
};

int main()
{
    ReverseString2 cls;
    cout << cls.reverseStr("abcdefg", 2).c_str() << " -> bacdfeg" << endl;
    cout << cls.reverseStr("abcd", 2).c_str() << " -> bacd" << endl;
    return 0;
}
